third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -415,9 +455,21 @@ PjRtComputationClient::ExecuteComputation(
                            returned_future)
           .value();
 
+  std::shared_ptr<PJRTLocker> locker = GetLockerForDevice(device);
```

Comment:
Would it be better to have the PjrtComputationClient to have a field with the map of shared pointers to the lockers, rather than a function-level static variable? Then this would be a member function call, and the storage would be the same as the classes storage stategy?

Reply:
Can we even just have a map of device strings to [`shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex)? Each execution can take a [`shared_lock`](https://en.cppreference.com/w/cpp/thread/shared_lock) and hold it until the callback finishes, and then you can implement `PJRTLocker::Barrier` by blocking on a `unique_lock` on the same `shared_mutex`